File: scripts/min_solve_network.jl
#!julia

using DataStructures, DataFrames, CSV, Base.Filesystem

fn, timefn, mproffn = ARGS

gurobi_options = Dict(:Method=>2, :Crossover=>0, :BarConvTol=>1e-5, :LogToConsole=>false, :Threads=>2)

times = OrderedDict()

let m
    times["import"] = @elapsed using EnergyModels, Gurobi
    times["loading"] = @elapsed m = EnergyModel(fn, solver=GurobiSolver(; gurobi_options...))
    times["building"] = @elapsed build(m)
    times["solving"] = @elapsed solve(m)
end # allow gc of m

let m
    times["loading 2"] = @elapsed m = EnergyModel(fn, solver=GurobiSolver(; gurobi_options...))
    times["building 2"] = @elapsed build(m)
    times["solving 2"] = @elapsed solve(m)
end

df = DataFrame(:step=>collect(keys(times)), :secs=>collect(values(times)))
CSV.write(timefn, df)

cp(first(filter(x->startswith(x, "mprofile"), readdir())), mproffn)

File: scripts/min_solve_network.py
#!python3

from vresutils.benchmark import timer
import pandas as pd
import sys
from glob import glob
from shutil import copyfile

fn, timefn, mproffn = sys.argv[1:4]

formulation ="kirchhoff"
solver = "gurobi"
solver_options = {"Method": 2, "Crossover": 0, "BarConvTol": 1e-5, 'LogToConsole': False, "Threads": 2}

times = pd.Series()

with timer("import modules") as t:
    import pypsa
times["import"] = t.usec

with timer("loading network") as t:
    n = pypsa.Network(fn)
times["loading"] = t.usec

with timer("building pyomo model") as t:
    pypsa.opf.network_lopf_build_model(n, formulation="kirchhoff")
times["building"] = t.usec

with timer("solving model") as t:
    with timer("prepare solver"):
        pypsa.opf.network_lopf_prepare_solver(n, solver)
    with timer("solving"):
        pypsa.opf.network_lopf_solve(n, formulation=formulation, solver_options=solver_options)
times["solving"] = t.usec

df = pd.DataFrame(dict(step=times.index, secs=times.values/1e6))
df.to_csv(timefn, index=False)

copyfile(glob("mprofile_*")[0], mproffn)
